2016/presentations/splash/text.md

```diff
@@ -0,0 +1,83 @@
+
+Why did we decide to create Splash? Main reason is that we needed some rendering for Scrapy. Scrapinghub
+was founded by founders of Scrapy - framework for creating web crawlers. 
+Today main challenge for all crawlers is JavaScript. If you want
+to know how important JavaScript is for modern web try disabling JavaScript in browser and try
+browsing your favorite webpages. I can guarantee you that in many cases you won't see anything. If
+you do manage to browse some content you'll see terribly broken layouts, html elements completely
+out of order, divs and headeings flying all over the place, missing data. In general you'll see
+chaos and mayhem. Long time ago people believed in a thing called "graceful degradation", meaning that 
+if user didnt have JavaScript page should still work but in a limited way. These days no one really
+does graceful degradation. If you dont have modern browser with JavaScript enabled you're not able
+to benefit from all the good content created by web authors. 
+
+Now you can of course bypass these limitations, usually web pages are doing plain HTTP requests
+in the background using AJAX, and AJAX is just normal HTTP request that you can imitate easily from spider.
+But most people dont know this or are lazy and dont want to spend hours of their time reverse engineering web
+pages by looking into developer tools, so they decide they prefer to render page. There are also some
+use cases when you really need to render JavaScript (e.g. you need screenshot of page or you need
+to deal with some bot protections that require you to solve JS challenge).
+
+Why we decided to create Splash and not go with other solution? In other words why not Selenium or 
+phantom JS? Main reason is that we really needed to have headless browser with HTTP API. We didnt
+want to add JavaScript rendering into Scrapy but thought that having separate service doing rendering
+will be better. So it was crucial to have HTTP API. None of existing JS rendering services provided
+this out of the box. Now you can probably add HTTP API to existing rendering frameworks, but once you 
+start doing this you realize it's hard work and it's not that easy to tie it existing software. On
+the other hand getting library for rendering web pages is not difficult in Python - look no further
+just use PyQt. In fact in our case rendering is done by Qt Web Kit. 
+
+Under the hood Splash architecture is simple. HTTP API part is written in Python Twisted. It makes
+heavy use of Twisted Deferred and Twisted.Resources. Main benefits it gives us is that we application
+is asynchronous which is absolutely crucial for this type of task. After request is received by 
+application it is forwarded to BrowserTab object. Every browser tab
+object gets a web page. Web page inherits from Qt QWebView. Web page does all the rendering, it 
+executes page JavaScript, renders the DOM and returns result. 
+
+How open source works in case of Splash?
```

Comment:
Same case as before. If this is a question then it should be:

How does open source work in case of Splash?

otherwise there should be no question mark at the end